File: frontend/src/app/services/shopping.ts
import { IngredientCategory, ShoppingItem, WeekPlan } from '../models/plan.model';

// Порядок категорий в списке покупок.
const CATEGORY_ORDER: IngredientCategory[] = [
  'Мясо и птица',
  'Рыба',
  'Овощи',
  'Молочное',
  'Бакалея',
  'Специи',
  'Прочее',
];

export interface ShoppingGroup {
  category: IngredientCategory;
  items: ShoppingItem[];
}

// Агрегирует ингредиенты всех блюд плана: суммирует по (name + unit),
// группирует по категориям в заданном порядке.
export function buildShoppingList(plan: WeekPlan): ShoppingGroup[] {
  const merged = new Map<string, ShoppingItem>();

  for (const dish of plan.dishes) {
    for (const ing of dish.ingredients) {
      const key = `${ing.name.toLowerCase()}__${ing.unit}`;
      const existing = merged.get(key);
      if (existing) {
        existing.qty += ing.qty;
      } else {
        merged.set(key, {
          name: ing.name,
          qty: ing.qty,
          unit: ing.unit,
          category: ing.category,
          checked: false,
        });
      }
    }
  }

  const byCategory = new Map<IngredientCategory, ShoppingItem[]>();
  for (const item of merged.values()) {
    const list = byCategory.get(item.category) ?? [];
    list.push(item);
    byCategory.set(item.category, list);
  }

  return CATEGORY_ORDER.filter((c) => byCategory.has(c)).map((category) => ({
    category,
    items: byCategory.get(category)!.sort((a, b) => a.name.localeCompare(b.name, 'ru')),
  }));
}

export function shoppingKey(item: ShoppingItem): string {
  return `${item.name.toLowerCase()}__${item.unit}`;
}

File: backend/app/services/shopping.py
from ..schemas import Dish, ShoppingGroup, ShoppingItem

CATEGORY_ORDER = [
    "Мясо и птица",
    "Рыба",
    "Овощи",
    "Молочное",
    "Бакалея",
    "Специи",
    "Прочее",
]


def build_shopping_list(dishes: list[Dish]) -> list[ShoppingGroup]:
    """Суммирует ингредиенты по (name+unit) и группирует по категориям."""
    merged: dict[str, ShoppingItem] = {}
    for dish in dishes:
        for ing in dish.ingredients:
            key = f"{ing.name.lower()}__{ing.unit}"
            if key in merged:
                merged[key].qty += ing.qty
            else:
                merged[key] = ShoppingItem(
                    name=ing.name, qty=ing.qty, unit=ing.unit, category=ing.category
                )

    by_cat: dict[str, list[ShoppingItem]] = {}
    for item in merged.values():
        by_cat.setdefault(item.category, []).append(item)

    order = CATEGORY_ORDER + [c for c in by_cat if c not in CATEGORY_ORDER]
    groups: list[ShoppingGroup] = []
    for cat in order:
        if cat in by_cat:
            items = sorted(by_cat[cat], key=lambda x: x.name.lower())
            groups.append(ShoppingGroup(category=cat, items=items))
    return groups
